airflow/models/dagbag.py

```diff
@@ -317,9 +317,7 @@ def kill_zombies(self, zombies, session=None):
         had a heartbeat for too long, in the current DagBag.
 
         :param zombies: zombie task instances to kill.
-        :type zombies: airflow.utils.dag_processing.SimpleTaskInstance
```

Comment:
That's what I tried to do, but when you start doing this, you uncover a whole host of further cyclic dependencies, which will take quite some time to untangle (like DagBag using TaskInstances which are using DAGs which are using ... DagBag). 

When you define those types in return or parameter values, you no longer can use local imports because the types become part of the public API of the class. And pretty much hell breaks loose. Some of these changes will require some more "internal architecture" changes - moving methods between tasks, separating out constants which are shared with logic, joining several separate but tightly coupled classes in single module because they currently refer each other. But doing it in this PR is not a good idea.

I already have a WIP where I added more types (https://github.com/PolideaInternal/airflow/tree/WIP-more-types) but this has so many changes that I decided to use it as a base to introduce more types later. There are a number of small things that we can do to untangle all of that (but I prefer to have steady stream of separate PRs for that) for example in order to avoid some cyclic dependencies between BaseOperator and TaskInstance we will have to move clear()/clear_task_instances() from BaseOperator (this actually makes sense and can be safely done but it's quite a change on it's own)

Reply:
It sounds like we should add this back, it's for docs not the type syste anyway.

The way to avoid imported modules/classes becoming part of the public API of this module is to use `__all__`